connmgr/banmanager.go

```diff
@@ -0,0 +1,202 @@
+		return false
+	}
+
+	for _, ipnet := range bm.cfg.WhiteList {
+		if ipnet.Contains(ip) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// AddPeer adds the provided peer to the ban manager.
+func (bm *BanManager) AddPeer(p *peer.Peer) {
+	bmp := &banMgrPeer{
+		Peer:          p,
+		isWhitelisted: bm.IsPeerWhitelisted(p),
+	}
+
+	bm.mtx.Lock()
+	bm.peers[p.Addr()] = bmp
+	bm.banScores[p.Addr()] = &DynamicBanScore{}
+	bm.mtx.Unlock()
+}
+
+// LookupPeer returns the associated peer of the provided address and its
+// ban score.
+func (bm *BanManager) LookupPeer(addr string) (*peer.Peer, uint32) {
+	bm.mtx.Lock()
+	p, ok := bm.peers[addr]
+	if !ok {
+		bm.mtx.Unlock()
+		return nil, 0
+	}
+	score := bm.banScores[addr]
+	bm.mtx.Unlock()
+
+	return p.Peer, score.Int()
+}
+
+// RemovePeer discards the provided peer from the ban manager.
+func (bm *BanManager) RemovePeer(p *peer.Peer) {
+	bm.mtx.Lock()
+	delete(bm.peers, p.Addr())
+	delete(bm.banScores, p.Addr())
+	bm.mtx.Unlock()
+}
+
+// BanPeer bans the provided peer.
+func (bm *BanManager) BanPeer(p *peer.Peer) {
+	// Return immediately if banning is disabled.
+	if bm.cfg.DisableBanning {
+		return
+	}
+
+	bm.mtx.Lock()
+	bmp, ok := bm.peers[p.Addr()]
+	if !ok {
+		bm.mtx.Unlock()
+		log.Errorf("No peer found with IP: '%s'", p.Addr())
+		return
+	}
+	bm.mtx.Unlock()
+
+	// Return if the peer is whitelisted.
+	if bmp.isWhitelisted {
+		return
+	}
+
+	// Ban and remove the peer.
+	bm.cfg.BanPeer(p)
+	bm.RemovePeer(p)
+}
+
+// AddBanScore increases the persistent and decaying ban scores of the
+// provided peer by the values passed as parameters. If the resulting score
+// exceeds half of the ban threshold, a warning is logged including the reason
+// provided. Further, if the score is above the ban threshold, the peer will
+// be banned.
+func (bm *BanManager) AddBanScore(p *peer.Peer, persistent, transient uint32, reason string) bool {
+	// No warning is logged and no score is calculated if banning is disabled.
+	if bm.cfg.DisableBanning {
+		return false
+	}
+
+	bm.mtx.Lock()
+	bmp, ok := bm.peers[p.Addr()]
+	if !ok {
+		bm.mtx.Unlock()
+		log.Debugf("No peer found with IP: '%s'", p.Addr())
+		return false
+	}
+	bm.mtx.Unlock()
+
+	if bmp.isWhitelisted {
+		log.Debugf("Misbehaving whitelisted peer %s: %s", p, reason)
+		return false
+	}
+
+	banScore, ok := bm.banScores[p.Addr()]
+	if !ok {
+		log.Debugf("No ban score found for peer %s: %s", p)
+		return false
+	}
+
+	warnThreshold := bm.cfg.BanThreshold >> 1
+	if transient == 0 && persistent == 0 {
+		// The score is not being increased, but a warning message is still
+		// logged if the score is above the warn threshold.
+		score := banScore.Int()
+		if score > warnThreshold {
+			log.Warnf("Misbehaving peer %s: %s -- ban score is %d, "+
+				"it was not increased this time", p, reason, score)
+		}
+		return false
+	}
+
+	score := banScore.Increase(persistent, transient)
+	bm.banScores[p.Addr()] = banScore
```

Comment:
No need to set it again.  So long as `banScore` is a pointer (as my previous comment set it to) `Increase` updates it.